File: evopanel/src/react/lib/chat-artifact.ts
/** Chat deliverable (产物) — shared by Info Rail + Right Stage. */

export type ChatArtifactType = 'file' | 'image' | 'video' | 'url' | 'html' | 'text' | 'platform'

export type ChatArtifact = {
  id: string
  type: ChatArtifactType
  name: string
  path?: string
  url?: string
  label?: string
  mime?: string
  size?: number
  status?: 'new' | 'updated'
  content?: string
  createdAt?: string
  updatedAt?: string
  /** platform tool success entry */
  platformAction?: string
  platformDomain?: string
  platformFeedbackKind?: 'success' | 'warning' | 'error'
  platformSubtitle?: string
  toolCallId?: string
  platformActions?: Array<{ label: string; route?: string; external?: boolean }>
}

const TYPES = new Set<ChatArtifactType>(['file', 'image', 'video', 'url', 'html', 'text', 'platform'])

function basename(pathOrUrl: string): string {
  const s = String(pathOrUrl || '').replace(/\\/g, '/').replace(/\/+$/, '')
  if (!s) return ''
  const parts = s.split('/').filter(Boolean)
  return parts.length ? parts[parts.length - 1] : s
}

function inferType(path: string, url: string, mime: string, rawType: string): ChatArtifactType {
  const t = String(rawType || '').trim().toLowerCase()
  if (TYPES.has(t as ChatArtifactType)) return t as ChatArtifactType
  const m = String(mime || '').toLowerCase()
  if (m.startsWith('image/')) return 'image'
  if (m.startsWith('video/')) return 'video'
  if (m.includes('html')) return 'html'
  const probe = (path || url).toLowerCase().split('?')[0]
  if (/\.(png|jpe?g|gif|webp|bmp|svg|ico)$/.test(probe)) return 'image'
  if (/\.(mp4|webm|mov|mkv|avi|m4v)$/.test(probe)) return 'video'
  if (/\.html?$/.test(probe)) return 'html'
  if (url) return 'url'
  return 'file'
}

/** Normalize API / SSE / panel item into ChatArtifact. */
export function normalizeChatArtifact(raw: unknown): ChatArtifact | null {
  if (raw == null) return null
  if (typeof raw === 'string') {
    const s = raw.trim()
    if (!s) return null
    if (/^https?:\/\//i.test(s)) return normalizeChatArtifact({ type: 'url', url: s })
    return normalizeChatArtifact({ type: 'file', path: s })
  }
  if (typeof raw !== 'object') return null
  const o = raw as Record<string, unknown>
  const typeRaw = String(o.type || o.kind || '').trim().toLowerCase()
  if (typeRaw === 'platform') {
    const label = String(o.label || o.title || o.name || '').trim()
    if (!label) return null
    const id = String(o.id || o.artifact_id || o.toolCallId || o.tool_call_id || label).trim()
    const route = String(o.url || o.route || '').trim()
    const statusRaw = String(o.status || '').trim().toLowerCase()
    const status = statusRaw === 'updated' ? 'updated' : statusRaw === 'new' ? 'new' : undefined
    const feedbackKindRaw = String(o.platformFeedbackKind || o.platform_feedback_kind || '').trim().toLowerCase()
    const platformFeedbackKind =
      feedbackKindRaw === 'warning' || feedbackKindRaw === 'error' ? feedbackKindRaw : 'success'
    const out: ChatArtifact = {
      id: id || `platform:${label}`,
      type: 'platform',
      name: label,
      label,
      platformFeedbackKind,
    }
    if (route) out.url = route
    if (status) out.status = status
    const platformAction = String(o.platformAction || o.platform_action || '').trim()
    const platformDomain = String(o.platformDomain || o.platform_domain || '').trim()
    const platformSubtitle = String(o.platformSubtitle || o.platform_subtitle || o.subtitle || '').trim()
    const toolCallId = String(o.toolCallId || o.tool_call_id || '').trim()
    if (platformAction) out.platformAction = platformAction
    if (platformDomain) out.platformDomain = platformDomain
    if (platformSubtitle) out.platformSubtitle = platformSubtitle
    if (toolCallId) out.toolCallId = toolCallId
    const createdAt = String(o.createdAt || o.created_at || '').trim()
    const updatedAt = String(o.updatedAt || o.updated_at || '').trim()
    if (createdAt) out.createdAt = createdAt
    if (updatedAt) out.updatedAt = updatedAt
    const actionsRaw = o.platformActions || o.platform_actions || o.actions
    if (Array.isArray(actionsRaw) && actionsRaw.length) {
      out.platformActions = actionsRaw
        .filter((row): row is Record<string, unknown> => !!row && typeof row === 'object')
        .map((row) => ({
          label: String(row.label || '').trim(),
          route: String(row.route || '').trim() || undefined,
          external: row.external === true,
        }))
        .filter((row) => row.label)
    }
    return out
  }
  const path = String(o.path || '').trim()
  const url = String(o.url || o.href || o.link || '').trim()
  const content = o.content != null ? String(o.content) : ''
  if (!path && !url && !content) return null
  const mime = String(o.mime || o.mime_type || '').trim()
  const type = inferType(path, url, mime, String(o.type || o.kind || ''))
  const name =
    String(o.name || o.filename || '').trim() ||
    basename(path || url) ||
    String(o.label || o.title || type)
  const id = String(o.id || o.artifact_id || path || url || name).trim() || name
  const statusRaw = String(o.status || '').trim().toLowerCase()
  const status = statusRaw === 'updated' ? 'updated' : statusRaw === 'new' ? 'new' : undefined
  const out: ChatArtifact = { id, type, name }
  if (path) out.path = path
  if (url) out.url = url
  if (content && (type === 'html' || type === 'text')) out.content = content
  const label = String(o.label || o.title || '').trim()
  if (label) out.label = label
  if (mime) out.mime = mime
  if (o.size != null && Number.isFinite(Number(o.size))) out.size = Math.max(0, Number(o.size))
  if (status) out.status = status
  const createdAt = String(o.createdAt || o.created_at || '').trim()
  const updatedAt = String(o.updatedAt || o.updated_at || '').trim()
  if (createdAt) out.createdAt = createdAt
  if (updatedAt) out.updatedAt = updatedAt
  return out
}

export function normalizeChatArtifacts(raw: unknown): ChatArtifact[] {
  const list = Array.isArray(raw) ? raw : raw == null ? [] : [raw]
  const out: ChatArtifact[] = []
  const seen = new Set<string>()
  for (const it of list) {
    const n = normalizeChatArtifact(it)
    if (!n || seen.has(n.id)) continue
    seen.add(n.id)
    out.push(n)
  }
  return out
}

/** Merge upsert into session list (append / replace by id). */
export function mergeChatArtifacts(
  existing: ChatArtifact[],
  incoming: ChatArtifact[],
): ChatArtifact[] {
  if (!incoming.length) return existing
  const now = new Date().toISOString()
  const map = new Map<string, ChatArtifact>()
  for (const it of existing) map.set(it.id, it)
  for (const raw of incoming) {
    const it = { ...raw }
    const prev = map.get(it.id)
    if (!prev) {
      if (!it.createdAt) it.createdAt = now
      if (!it.updatedAt) it.updatedAt = now
    } else if (!it.updatedAt) {
      it.updatedAt = now
    }
    map.set(it.id, prev ? { ...prev, ...it, status: it.status || prev.status || 'updated' } : it)
  }
  // Keep prior order; append brand-new ids at end
  const order: string[] = []
  const seen = new Set<string>()
  for (const it of existing) {
    if (map.has(it.id) && !seen.has(it.id)) {
      order.push(it.id)
      seen.add(it.id)
    }
  }
  for (const it of incoming) {
    if (!seen.has(it.id)) {
      order.push(it.id)
      seen.add(it.id)
    }
  }
  return order.map((id) => map.get(id)!).filter(Boolean)
}

export function chatArtifactDisplayLabel(it: ChatArtifact): string {
  if (it.type === 'platform') {
    return String(it.label || it.name || it.platformAction || '平台操作').trim()
  }
  return String(it.label || it.name || it.path || it.url || it.id || '产物').trim()
}

export function chatArtifactTypeLabel(it: ChatArtifact): string {
  if (it.type === 'platform') {
    const domain = String(it.platformDomain || '').trim()
    const map: Record<string, string> = {
      items: '待办',
      tasks: '协作任务',
      employees: '员工',
      agents: '智能体',
      skills: '技能',
      mcp: 'MCP',
      automation: '自动化',
      settings: '设置',
      knowledge: '知识库',
    }
    return domain ? `平台 · ${map[domain] || domain}` : '平台'
  }
  return it.type
}

File: backend/packages/harness/evoflow/artifacts/chat_artifact.py
"""Chat deliverable (产物) contract — structured present, not @@ scrape.

Types: file | image | video | url | html | text
Location: path XOR url XOR content (at least one).
"""

from __future__ import annotations

import hashlib
import re
from typing import Any

ARTIFACT_TYPES = frozenset({"file", "image", "video", "url", "html", "text", "platform"})

_IMAGE_EXT = frozenset({".png", ".jpg", ".jpeg", ".gif", ".webp", ".bmp", ".svg", ".ico"})
_VIDEO_EXT = frozenset({".mp4", ".webm", ".mov", ".mkv", ".avi", ".m4v"})
_HTML_EXT = frozenset({".html", ".htm"})

_NAME_MAX = 240
_LABEL_MAX = 200
_PATH_MAX = 2000
_URL_MAX = 4000
_CONTENT_MAX = 200_000
_ID_MAX = 128


def normalize_artifact_type(raw: Any, *, path: str = "", url: str = "", mime: str = "") -> str:
    t = str(raw or "").strip().lower()
    if t in ARTIFACT_TYPES:
        return t
    if t in {"path", "filepath", "doc", "document", "attachment"}:
        return "file"
    if t in {"link", "uri", "href"}:
        return "url"
    if t in {"note", "string", "markdown", "md"}:
        return "text"
    if t in {"webpage", "page"}:
        return "html"
    mime_l = str(mime or "").strip().lower()
    if mime_l.startswith("image/"):
        return "image"
    if mime_l.startswith("video/"):
        return "video"
    if mime_l in {"text/html", "application/xhtml+xml"}:
        return "html"
    probe = (path or url or "").lower().split("?", 1)[0]
    for ext in _IMAGE_EXT:
        if probe.endswith(ext):
            return "image"
    for ext in _VIDEO_EXT:
        if probe.endswith(ext):
            return "video"
    for ext in _HTML_EXT:
        if probe.endswith(ext):
            return "html"
    if str(url or "").strip():
        return "url"
    return "file"


def _basename(path_or_url: str) -> str:
    s = str(path_or_url or "").replace("\\", "/").rstrip("/")
    if not s:
        return ""
    return s.rsplit("/", 1)[-1] or s


def make_artifact_id(*, type_: str, path: str = "", url: str = "", content: str = "", name: str = "") -> str:
    """Stable identity for append/upsert (not a display label)."""
    if path:
        key = f"path:{path.replace(chr(92), '/')}"
    elif url:
        key = f"url:{url}"
    elif content:
        digest = hashlib.sha256(content.encode("utf-8", errors="ignore")).hexdigest()[:24]
        key = f"content:{type_}:{name or 'untitled'}:{digest}"
    else:
        key = f"empty:{type_}:{name or 'untitled'}"
    # Keep PK-friendly length
    if len(key) <= _ID_MAX:
        return key
    digest = hashlib.sha256(key.encode("utf-8")).hexdigest()[:40]
    return f"h:{digest}"


def normalize_chat_artifact(raw: Any) -> dict[str, Any] | None:
    """Normalize one tool/API payload into a ChatArtifact dict, or None if invalid."""
    if raw is None:
        return None
    if isinstance(raw, str):
        s = raw.strip()
        if not s:
            return None
        if re.match(r"^https?://", s, re.I):
            return normalize_chat_artifact({"type": "url", "url": s})
        return normalize_chat_artifact({"type": "file", "path": s})

    if not isinstance(raw, dict):
        return None

    type_raw = str(raw.get("type") or raw.get("kind") or "").strip().lower()
    if type_raw == "platform":
        title = str(raw.get("label") or raw.get("title") or raw.get("name") or "").strip()[:_LABEL_MAX]
        if not title:
            return None
        tcid = str(raw.get("toolCallId") or raw.get("tool_call_id") or "").strip()
        action = str(raw.get("platformAction") or raw.get("platform_action") or raw.get("action") or "").strip()
        entity_id = str(raw.get("entityId") or raw.get("entity_id") or "").strip()
        aid = str(raw.get("id") or raw.get("artifact_id") or "").strip()
        if not aid:
            if tcid:
                aid = f"platform:{tcid}"
            elif action:
                aid = f"platform:{action}:{entity_id or title}"
            else:
                aid = f"platform:{title}"
        route = str(raw.get("url") or raw.get("route") or "").strip()[:_URL_MAX]
        kind_raw = str(raw.get("platformFeedbackKind") or raw.get("platform_feedback_kind") or raw.get("kind") or "").strip().lower()
        kind = kind_raw if kind_raw in {"warning", "error"} else "success"
        out: dict[str, Any] = {
            "id": aid[:_ID_MAX],
            "type": "platform",
            "name": title,
            "label": title,
            "platformFeedbackKind": kind,
        }
        if route:
            out["url"] = route
        if action:
            out["platformAction"] = action
        domain = str(raw.get("platformDomain") or raw.get("platform_domain") or raw.get("domain") or "").strip()
        if domain:
            out["platformDomain"] = domain
        subtitle = str(raw.get("platformSubtitle") or raw.get("platform_subtitle") or raw.get("subtitle") or "").strip()
        if subtitle:
            out["platformSubtitle"] = subtitle
        if tcid:
            out["toolCallId"] = tcid
        actions = raw.get("platformActions") or raw.get("platform_actions") or raw.get("actions")
        if isinstance(actions, list) and actions:
            out["platformActions"] = actions
        status = str(raw.get("status") or "").strip().lower()
        if status in {"new", "updated"}:
            out["status"] = status
        return out

    path = str(raw.get("path") or "").strip()[:_PATH_MAX]
    url = str(raw.get("url") or raw.get("href") or raw.get("link") or "").strip()[:_URL_MAX]
    content = raw.get("content")
    content_s = "" if content is None else str(content)[:_CONTENT_MAX]
    mime = str(raw.get("mime") or raw.get("mime_type") or "").strip()[:120]
    name = str(raw.get("name") or raw.get("filename") or "").strip()[:_NAME_MAX]
    label = str(raw.get("label") or raw.get("title") or "").strip()[:_LABEL_MAX]
    type_ = normalize_artifact_type(raw.get("type") or raw.get("kind"), path=path, url=url, mime=mime)

    if not path and not url and not content_s:
        return None

    if not name:
        name = _basename(path or url) or (label if label else type_)

    aid = str(raw.get("id") or raw.get("artifact_id") or "").strip()
    if not aid:
        aid = make_artifact_id(type_=type_, path=path, url=url, content=content_s, name=name)

    size_raw = raw.get("size")
    size: int | None = None
    if size_raw is not None:
        try:
            size = max(0, int(size_raw))
        except (TypeError, ValueError):
            size = None

    out: dict[str, Any] = {
        "id": aid[:_ID_MAX],
        "type": type_,
        "name": name,
    }
    if path:
        out["path"] = path
    if url:
        out["url"] = url
    if content_s and type_ in {"html", "text"}:
        out["content"] = content_s
    if label:
        out["label"] = label
    if mime:
        out["mime"] = mime
    if size is not None:
        out["size"] = size
    status = str(raw.get("status") or "").strip().lower()
    if status in {"new", "updated"}:
        out["status"] = status
    return out


def normalize_chat_artifacts(raw: Any, *, limit: int = 40) -> list[dict[str, Any]]:
    if raw is None:
        return []
    items: list[Any]
    if isinstance(raw, list):
        items = raw
    else:
        items = [raw]
    out: list[dict[str, Any]] = []
    seen: set[str] = set()
    for it in items:
        if len(out) >= limit:
            break
        n = normalize_chat_artifact(it)
        if not n:
            continue
        aid = str(n["id"])
        if aid in seen:
            continue
        seen.add(aid)
        out.append(n)
    return out


def artifact_state_key(item: dict[str, Any]) -> str:
    """ThreadState.artifacts still stores string keys (path/url/id)."""
    path = str(item.get("path") or "").strip()
    if path:
        return path
    url = str(item.get("url") or "").strip()
    if url:
        return url
    return str(item.get("id") or "").strip()
